base/multidimensional.jl

```diff
@@ -616,33 +615,45 @@ module IteratorsMD
         # In general, the Cartesian Partition might start and stop in the middle of the outer
         # dimensions — thus the outer range of a CartesianPartition is itself a
         # CartesianPartition.
-        t = tail(iter.parent.parent.indices)
-        ci = CartesianIndices(t)
-        li = LinearIndices(t)
-        return @inbounds view(ci, li[tail(iter[1].I)...]:li[tail(iter[end].I)...])
+        mi = iter.parent.mi
+        ci = iter.parent.parent
+        ax, ax1 = axes(ci), Base.axes1(ci)
+        subs = Base.ind2sub_rs(ax, mi, first(iter.indices[1]))
+        vl, fl = Base._sub2ind(tail(ax), tail(subs)...), subs[1]
+        vr, fr = divrem(last(iter.indices[1]) - 1, mi[end]) .+ (1, first(ax1))
+        oci = CartesianIndices(tail(ci.indices))
+        # A fake CartesianPartition to reuse the outer iterate fallback
+        outer = @inbounds view(ReshapedArray(oci, (length(oci),), mi), vl:vr)
```

Comment:
All test passed except for freebsd64, so I think this PR is ready for further review. @vtjnash @mbauman
Since `ReshapedArray` use stride based `mi` to accelerate index transformation, I have to form the `simd_outer_range` with an illegal `CartesianPartition`. If `mi` carries size infomation, then we can reuse the input's `mi` safely, and there's no need to use `Iterators.rest`. 

Previous commit shows that official test passed with size based `mi` (but pkgs might not). My local bench shows that size based `mi` could accelerate the sequential access of a ReshapedArray a little.
@timholy Is there a strong reason for us to use stride based `mi`? And is it acceptable to make `mi` carry size infomation in this PR?